File: MovieTicketBooking/Movie Booking Java/src/main/java/org/example/service/BookingService.java
package org.example.service;

import lombok.NonNull;
import org.example.exceptions.BadRequestException;
import org.example.exceptions.NotFoundException;
import org.example.exceptions.SeatAlreadyBookedException;
import org.example.model.Booking;
import org.example.model.Seat;
import org.example.model.Show;
import org.example.providers.SeatLockProvider;

import java.util.*;
import java.util.stream.Collectors;

public class BookingService {

    private final Map<String, Booking> bookings;
    private final SeatLockProvider seatLockProvider;

    public BookingService(SeatLockProvider seatLockProvider) {

        this.bookings = new HashMap<>();
        this.seatLockProvider = seatLockProvider;
    }

    public Booking getBooking(String id) {

        if (!bookings.containsKey(id)) {
            throw new NotFoundException();
        }

        return bookings.get(id);
    }

    public Booking createBooking(@NonNull final String userId,
                                 @NonNull final Show show,
                                 @NonNull final List<Seat> seats) {

        if (isAnySeatAlreadyBooked(show, seats)) {
            throw new SeatAlreadyBookedException();
        }

        String id = UUID.randomUUID().toString();
        Booking booking = new Booking(id, seats, userId, show);
        bookings.put(id, booking);
        seatLockProvider.lockSeat(seats, show, userId);
        return booking;
    }

    public List<Booking> getAllBookingsForShow(@NonNull final Show show) {

        List<Booking> response = new ArrayList<>();
        for (Booking booking : bookings.values()) {

            if (booking.getShow().equals(show)) {
                response.add(booking);
            }
        }

        return response;
    }

    public List<Seat> getBookedSeats(@NonNull final Show show) {

        return this.getAllBookingsForShow(show)
                .stream()
                .filter(Booking::isConfirmed)
                .map(Booking::getSeats)
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
    }

    public void confirmBooking(@NonNull final  Booking booking, @NonNull final String userId) {

        if (!booking.getUserId().equals(userId)) {
            throw new BadRequestException();
        }

        for (Seat seat : booking.getSeats()) {
            boolean isLockValidated = seatLockProvider.validateSeatLock(seat, booking.getShow(), userId);
            if (!isLockValidated) {
                throw new SeatAlreadyBookedException();
            }
        }
        booking.confirmBooking();
    }

    private boolean isAnySeatAlreadyBooked(Show show, List<Seat> seats){

        List<Seat> bookedSeats = getBookedSeats(show);

        for (Seat seat : seats) {

            if (bookedSeats.contains(seat)) {
                return true;
            }
        }

        return false;
    }

}

File: MovieTicketBooking/Movie Booking C#/Services/BookingService.cs
using MovieTicketBooking.Model;
using MovieTicketBooking.Providers;

namespace MovieTicketBooking.Services
{
    public class BookingService
    {
        ISeatLockProvider _seatLockProvider;
        Dictionary<string, Booking> _bookings = new();
        public BookingService(ISeatLockProvider seatLockProvider)
        {
            _seatLockProvider = seatLockProvider;
        }
        public string CreateBooking(Show show, string user, List<Seat> seats)
        {
            if(isAnySeatAlreadyBooked(show, seats))
            {
                throw new Exception("seats are booked");
            }
            _seatLockProvider.LockSeats(show, seats, user);
            string id = Guid.NewGuid().ToString();
            Booking booking = new Booking(id, BookingStatus.Created, user, show, seats);
            _bookings.Add(id, booking);
            return id;
        }
        public void ConfrimBooking(string bookingId, string user)
        {
            if(!_bookings.ContainsKey(bookingId))
            {
                throw new Exception("Booking does not exists");
            }
            Booking booking = _bookings.GetValueOrDefault(bookingId);
            List<Seat> bookedSeats = booking.Seats;
            foreach(Seat seat in bookedSeats)
            {
                if(!_seatLockProvider.ValidateLock(booking.Show, seat, user))
                {
                    throw new Exception("Lock is not applied on seat with id "+  seat.Id);
                }
            }
            booking.Status = BookingStatus.Confirmed;
        } 
        private bool isAnySeatAlreadyBooked(Show show, List<Seat> seats)
        {
            List<Seat> bookedSeats = GetBookedSeats(show);
            foreach(Seat seat in seats)
            {
                if(bookedSeats.Contains(seat))
                {
                    return true;
                }
            }
            return false;
        }
        public List<Seat> GetBookedSeats(Show show)
        {
            List<Seat> seats = new();
            foreach(Booking booking in _bookings.Values)
            {
                if(booking.Show == show && booking.Status == BookingStatus.Confirmed)
                {
                    seats = booking.Seats;
                    return seats;
                }
            }
            return seats;
        }
        public Booking GetBooking(string id)
        {
            if(!_bookings.ContainsKey(id))
            {
                throw new Exception("Booking does not exists");
            }
            return _bookings.GetValueOrDefault(id);
        }
    }
}